pkg/readers/bufferedfilereader.go

```diff
@@ -0,0 +1,88 @@
+package readers
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+	bufferedfilewriter "github.com/trufflesecurity/trufflehog/v3/pkg/writers/buffered_file_writer"
+)
+
+// bufferedFileReader provides random access read, seek, and close capabilities on top of the BufferedFileWriter.
+// It combines the functionality of BufferedFileWriter for buffered writing, io.ReadSeekCloser for
+// random access reading and seeking.
+type bufferedFileReader struct {
+	bufWriter *bufferedfilewriter.BufferedFileWriter
+	reader    io.ReadSeekCloser
+}
+
+// NewBufferedFileReader initializes a bufferedFileReader from an io.Reader by using
+// the BufferedFileWriter's functionality to read and store data, then setting up a io.ReadSeekCloser
+// for random access reading and seeking.
+// Close should be called when the bufferedFileReader is no longer needed.
+// It returns the initialized bufferedFileReader and any error encountered during the process.
+func NewBufferedFileReader(ctx context.Context, r io.Reader) (*bufferedFileReader, error) {
+	writer, err := bufferedfilewriter.NewFromReader(ctx, r)
+	if err != nil {
+		return nil, fmt.Errorf("error creating bufferedFileReader: %w", err)
+	}
+
+	// Ensure that the BufferedFileWriter is in read-only mode.
+	if err := writer.CloseForWriting(); err != nil {
+		return nil, err
+	}
+
+	reader, err := writer.ReadCloser()
+	if err != nil {
+		return nil, err
+	}
+
+	rdr, ok := reader.(io.ReadSeekCloser)
```

Comment:
Since we control all the relevant implementations, is it correct to say that we always expect this to succeed? And if so, should `ReadCloser()` be changed to return a `*io.ReadSeekCloser` so that we don't have to do this at runtime?

Reply:
I considered raising this question in the pr: If we modify the return type to io.ReadSeekCloser, we will also need to revise the ReadCloser method in the contentReader interface within gitparse.go. This adjustment left me somewhat uncertain, but I'm happy to make the change.

I also tend to get a little squimish around interface assertions. 😅